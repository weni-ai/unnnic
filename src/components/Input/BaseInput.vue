<template>
  <input
    v-if="mask"
    v-mask="mask"
    v-bind="attributes"
    :value="fullySanitize(modelValue)"
    :class="[classes, { 'focus': forceActiveStatus }]"
    :type="nativeType"
    :readonly="readonly"
  />
  <input
    v-else
    v-bind="attributes"
    :value="fullySanitize(modelValue)"
    :class="[classes, { 'focus': forceActiveStatus }]"
    :type="nativeType"
    :maxlength="maxlength"
    :readonly="readonly"
  />
</template>

<script>
import { mask } from 'vue-the-mask';
import { fullySanitize } from '../../utils/sanitize';

export default {
  directives: { mask },
  props: {
    type: {
      type: String,
      default: 'normal',
      validator(value) {
        return ['normal', 'error'].indexOf(value) !== -1;
      },
    },
    modelValue: {
      type: String,
      default: '',
    },
    nativeType: {
      type: String,
      default: 'text',
    },
    size: {
      type: String,
      default: 'md',
    },
    mask: {
      type: [String, Array],
      default: '',
    },
    hasIconLeft: Boolean,
    hasIconRight: Boolean,
    hasClearIcon: Boolean,
    maxlength: {
      type: Number,
      default: null,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    forceActiveStatus: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['update:modelValue'],
  data() {
    return {};
  },
  computed: {
    attributes() {
      return {
        ...this.$attrs,
        ...this.$attrs['v-bind'],
        onInput: (event) => {
          this.$emit('update:modelValue', event.srcElement.value);
        },
      };
    },

    classes() {
      return [
        'input',
        `size-${this.size}`,
        this.type,
        {
          'input--has-icon-left': this.hasIconLeft,
          'input--has-icon-right': this.hasIconRight,
          'input--has-clear-icon': this.hasClearIcon,
        },
      ];
    },
  },
  methods: {
    fullySanitize,
  },
};
</script>

<style lang="scss" scoped>
@use '@/assets/scss/unnnic' as *;
@use './Input.scss' as *;

.input {
  @include input-base;

  box-sizing: border-box;
  width: 100%;

  &.size {
    &-md {
      @include input-md-font;
      padding: $unnnic-space-3 $unnnic-space-4;
    }

    &-sm {
      @include input-sm-font;
      padding: $unnnic-space-2 $unnnic-space-4;
    }
  }

  &.input--has-icon-left {
    padding-left: $unnnic-space-10;
  }

  &.input--has-icon-right {
    padding-right: $unnnic-space-10;
    &.input--has-clear-icon {
      padding-right: $unnnic-space-10 + $unnnic-space-6;
    }
  }

  &.input--has-clear-icon {
      padding-right: $unnnic-space-10;
    }

  &.error {
    @include input-error;
  }

  &:-ms-input-placeholder,
  &::-ms-input-placeholder {
    /* Internet Explorer 10-11 */
    color: $unnnic-color-fg-muted;
  }
}
</style>
