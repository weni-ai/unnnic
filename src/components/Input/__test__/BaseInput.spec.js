import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, afterEach, test, vi } from 'vitest';
import BaseInput from '@/components/Input/BaseInput.vue';
import { mask } from 'vue-the-mask';

describe('BaseInput.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(BaseInput, {
      props: {
        type: 'normal',
        modelValue: '',
        nativeType: 'text',
        size: 'md',
        mask: '',
      },
      global: {
        directives: {
          mask,
        },
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
  });

  test('applies the correct classes', async () => {
    expect(wrapper.classes()).toContain('input');
    expect(wrapper.classes()).toContain('size-md');
    expect(wrapper.classes()).toContain('normal');

    await wrapper.setProps({ type: 'error', size: 'sm' });
    expect(wrapper.classes()).toContain('size-sm');
    expect(wrapper.classes()).toContain('error');
  });

  test('renders input element with correct attributes and value', async () => {
    const input = wrapper.find('input');
    expect(input.exists()).toBe(true);
    expect(input.attributes('type')).toBe('text');
    expect(input.element.value).toBe('');

    await wrapper.setProps({
      modelValue: 'test value',
      nativeType: 'password',
    });
    expect(input.element.value).toBe('test value');
    expect(input.attributes('type')).toBe('password');
  });

  test('emits update:modelValue on input', async () => {
    const input = wrapper.find('input');
    await input.setValue('new value');

    expect(wrapper.emitted('update:modelValue')).toBeTruthy();
    expect(wrapper.emitted('update:modelValue')[0]).toEqual(['new value']);
  });

  test('applies the correct mask if provided', async () => {
    const maskedWrapper = mount(BaseInput, {
      props: {
        mask: '####-####',
        modelValue: '12345678',
      },
    });

    const input = maskedWrapper.find('input');

    expect(input.element.value).toBe('1234-5678');

    maskedWrapper.unmount();
  });

  test('does not apply mask if mask prop is empty', () => {
    const input = wrapper.find('input');
    expect(input.attributes('v-mask')).toBeUndefined();
  });

  test('focuses on input when focus method is called', () => {
    const input = wrapper.find('input');
    const focusSpy = vi.spyOn(input.element, 'focus');

    input.element.focus();

    expect(focusSpy).toHaveBeenCalled();
    focusSpy.mockRestore();
  });

  test('updates the attributes computed property correctly', async () => {
    await wrapper.setProps({ placeholder: 'Enter text', disabled: true });
    const attributes = wrapper.vm.attributes;

    expect(attributes.placeholder).toBe('Enter text');
    expect(attributes.disabled).toBe(true);
  });

  test('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
