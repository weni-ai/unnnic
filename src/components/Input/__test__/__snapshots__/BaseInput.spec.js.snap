// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseInput.vue > matches the snapshot 1`] = `"<input data-v-86533b41="" class="input size-md normal" type="text" value="">"`;
